Data/ByteString.hs

```diff
@@ -1471,16 +1490,18 @@ zipWith' :: (Word8 -> Word8 -> Word8) -> ByteString -> ByteString -> ByteString
 zipWith' f (BS fp l) (BS fq m) = unsafeDupablePerformIO $
     withForeignPtr fp $ \a ->
     withForeignPtr fq $ \b ->
-    create len $ zipWith_ 0 a b
+    create len $ go a b
   where
-    zipWith_ :: Int -> Ptr Word8 -> Ptr Word8 -> Ptr Word8 -> IO ()
-    zipWith_ !n !p1 !p2 !r
-       | n >= len = return ()
-       | otherwise = do
-            x <- peekByteOff p1 n
-            y <- peekByteOff p2 n
-            pokeByteOff r n (f x y)
-            zipWith_ (n+1) p1 p2 r
+    go p1 p2 = zipWith_ 0
```

Comment:
A general question.  Under what conditions is it actually better to avoid passing fixed arguments on the stack in recursive calls, and instead allocate a closure around the fixed arguments, that avoids having to copy them into each stack frame?

I must admit I don't really understand what the compiler does here, but if the new version has to perform closure allocations that the old version does not, then perhaps the cost of carrying the extra arguments around is worth the ultimate reduction in GC pressure.

If the benchmarks only measure speed, but don't measure GC pressure, we may not be doing the right thing here (and elsewhere...).  The benchmarks would have to run long enough to take the GC overhead into account...

Reply:
I think it is always preferable to avoid passing fixed arguments (unless it is a particularly degenerate case). An additional closure (if any) is allocated only once per call, which is not a big deal at all, but additional arguments are passed for each element of the string.

Reply:
Closure allocation is however much more expensive than pushing a parameter onto the stack, so for sufficiently short strings the savings may be small, and the ultimate GC costs non-negligible if one processes a lot of short strings, or even a lot of strings period, sometimes it is better to pay the price as you go than to accumulate a lot of garbage to later clean up and stall.

Mind you, I'd like to hear from @hsyl20 et. al. on what the actual tradeoffs are here.  I don't yet have a clear enough sense of when the various costs and benefits come into play...

Reply:
Yes closure allocation can be costly. I have a [patch for containers](https://gitlab.haskell.org/ghc/ghc/-/issues/18541#note_292432) that removes this kind of closure for a lookup in IntMap.

In this specific `zipWith'` case, we don't allocate any closure as far as I can tell from looking at STG. `create` and `zipWith_` are inlined and transformed into join points.